docs/proposals/vip-3.md

```diff
@@ -0,0 +1,226 @@
+
+### Why JNI?
+
+Because it is supported in Rust, and that should make the integration easier. An alternative would be to have the Java 
+server communicate with the Rust server via some other mechanism (loopback socket, pipes, etc) but then that would mean
+adding lots of forks in the code, and might also be less efficient. Using a non-JNI mechanism would also mean that just
+enabling the Rust server (even if it's not used yet) would already alter the behavior of the main code, thus creating
+more risk. In order to work in an incremental manner, where the Rust code can be developed in small chunks and hidden
+behind a config flag until ready, it is preferable to keep this interface the same.
+
+In the end state, JNI would only be used by the Java code responsible for the ingestion, and the read path would be
+fully JNI-free. Also, it should be noted that the Rust server will not be a separate child process spawned by the Java
+process, but rather fully embedded in the Java process. The Rust server will be activated by a JNI call initiated by the
+Java server, which spins up the gRPC Handler on the Rust side, along with any port and threads it needs to function.
+
+Doing JNI properly requires packaging a jar with code compiled for multiple architectures (at least Mac and Linux for
+the time being). There is some complexity there, but it can be done.
+
+### Quota
+
+The read quota will need to be re-implemented in Rust. There are essentially two main approaches we can consider:
+
+1. Let the Rust server and Java server share a quota amount between them, so that no matter if the traffic comes in via
+   the Java port or the Rust port, it will be accounted for and the configured limit will be respected. This requires
+   having some kind of synchronization between the two sides (either at the granularity of each request, which is likely
+   too expensive, or by reserving some allotment periodically, which is complicated).
+2. Let the Rust server and Java server each have the full quota amount and not need any interaction between each other 
+   for sharing it. This is much simpler to implement, but the downside is that if clients were to perfectly balance
+   their traffic across the two ports, they could get up to twice their configured quota.
+
+Since the migration is expected to be controlled by the operator, and should only be activated for users of the Fast
+Client (i.e. not for users that run a mix of Thin Clients and Fast Clients), the edge case solved by the more 
+complicated solution 1 does not seem likely to come up. Therefore, the proposal is to pick the simpler solution 2.
+
+Also, it can be considered whether quota should be part of the MVP at all. If the Rust server mode will be tried in 
+dedicated clusters at first, then quota may not even be a must-have (though of course it will be needed eventually).
+
+### Cluster Discovery
```

Comment:
ACL's also needs some retweak.

Reply:
Yeah. I mentioned it in passing in the Development Milestones section. I don't have much thoughts on how to flesh it out further for the time being, besides saying that it's needed. If you have any inputs, please share!

Reply:
Weeeeelllll, since you asked :)

Something I've always lamented about the current implementation is that for every single request we do a lookup against the validation library to determine if an action is valid.  This is a reflection on how the interface for our internal authorizer works.

In my view, there are a few problems with this performance wise.  The first one is Venice specific, the router only allows 1 kind of operation to a store, so it seems like a pointless lookup for every single request.  The second reason, is that it's not as if the principal changes on a connection.  It's always the same actor on that connection, but we still interrogate it's permissions on every request on that connection.  Add to it, the LinkedIn representation of an Ace is Resource->[Permissions], so every time we're walking a (albeit usually short) list of principals to match up to the principal making the request.  There can be multiple aces for the same principal in this array, so every time we walk the entire array.  Ideally it should be the opposite at query time, being indexed by the principal as opposed to the resource (basically Principal->listOfResources)

So, what might be doable, is that we negotiate the permission once per connection, at time of establishing the connection.  We could just call into the java interface we have today with 'hasAccess', check back on the result, and use that for weather or not we accept the connection or not, and then let the client do whatever they want.  We'd need to tweak the protocol a bit to pass along at time of establishing the store(s) they want to talk to. Also today the transport client doing connection management can be shared between multiple Venice clients, so we'd need to do some wiring in here.

Other complexity gotcha, for internal auditing we might need to tweak this behavior for LDAP typed certificates so it gets audited correctly, but thats an internal detail.

It seems to me that this would be the sweet spot in terms of performance and not have to do anything too drastic in terms of writing equivalent interface/implementations between the native and java layer, and we could even incrementally approach going full native provided that at some point down the road we get a good Rust implementation of the LinkedIn authorizer.

Reply:
Thanks! I think these are really good ideas!

So let me recap to see if I got it:

The Rust server would maintain a mapping of `{principal, resource} -> {isAuthorized, lastCheckTime}`. If the mapping is missing, it would call into the Java server in a blocking way to find out whether this principal is allowed to access that resource, and then update the mapping accordingly. Some background thread could refresh the mapping when `lastCheckTime` is beyond some threshold, which also serves as a heartbeat into the auditing on the Java side. In this way, a Venice operator only needs to implement the Authorizer API once, in Java, regardless of whether the Rust server is enabled or not.

Reply:
Thats about right.

Reply:
I've included this within the VIP. Thanks!